### PR TITLE
add Open Election data to JSON protocol + handler

### PR DESCRIPTION
**Pull request by ahominal**
- add open election message to JSON protocol
- add Open Election data + adding it to protocol
- add Open Election handler (should open election upon receiving the OpenElection message)
- add test to handler + open election

Next pull request (if this accepted) should add the process of sending the OpenElection

UPDATE:
- modify election action button for the organizer so he can starts election whenever he wants
- election upon start time automatically send open election message
- update code according to suggestions mage

### CODE REVIEW

**Review by Tesa27 — DISMISSED** [review]
Looks good to me.

**Review by jbsv — CHANGES_REQUESTED** [review]

File: fe2-android/app/src/main/java/com/github/dedis/popstellar/model/network/method/message/data/election/OpenElection.java [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+      return true;
+    }
+    if (o == null || getClass() != o.getClass()) {
+      return false;
+    }
+    OpenElection that = (OpenElection) o;
+
+    return openedAt == that.openedAt
+        && java.util.Objects.equals(laoId, that.laoId)
+        && java.util.Objects.equals(electionId, that.electionId);
```
[the remark this answers is not in the document]
Comment (ahominal) [reply fe2-android/app/src/main/java/com/github/dedis/popstellar/model/network/method/message/data/election/OpenElection.java:61]: Yes that looks better !

File: fe2-android/app/src/main/java/com/github/dedis/popstellar/utility/handler/data/ElectionHandler.java [hunk 2]
```diff
@@ -1,6 +1,7 @@
 package com.github.dedis.popstellar.utility.handler.data;
 
 import static com.github.dedis.popstellar.model.objects.event.EventState.CLOSED;
+import static com.github.dedis.popstellar.model.objects.event.EventState.CREATED;
```
[the remark this answers is not in the document]
Comment (ahominal) [reply fe2-android/app/src/main/java/com/github/dedis/popstellar/utility/handler/data/ElectionHandler.java:4]: I already corrected this on my next pull request, setup should set it to created.

File: fe2-android/app/src/test/java/com/github/dedis/popstellar/utility/handler/ElectionHandlerTest.java [hunk 3]
```diff
@@ -199,4 +204,29 @@ public void testHandleElectionEnd() throws DataHandlingException {
...
+
+  @Test
+  public void handleOpenElectionTest() throws DataHandlingException {
+    OpenElection openElection = new OpenElection(lao.getId(), election.getId(), openedAt);
+    MessageGeneral message = new MessageGeneral(SENDER_KEY, openElection, GSON);
+
+    for (EventState state : EventState.values()) {
+      election.setEventState(state);
+
+      if (state == EventState.CREATED) {
```
[the remark this answers is not in the document]
Comment (ahominal) [reply fe2-android/app/src/test/java/com/github/dedis/popstellar/utility/handler/ElectionHandlerTest.java:216]: After it handles the open election message, it should switch to CREATED --> OPENED? this is what i test

**Review by zoemarin — COMMENTED** [review]
Also looks fine for me 
I agree to the class corresponding to the json

File: fe2-android/app/src/main/java/com/github/dedis/popstellar/model/network/method/message/data/election/OpenElection.java [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+import com.github.dedis.popstellar.model.network.method.message.data.Objects;
+import com.google.gson.annotations.SerializedName;
+
+public final class OpenElection extends Data {
+
+  @SerializedName("lao")
+  private final String laoId;
+
+  @SerializedName("election")
+  private final String electionId;
```
Comment (GabrielFleischer) [inline fe2-android/app/src/main/java/com/github/dedis/popstellar/model/network/method/message/data/election/OpenElection.java:19]: I see that you're using `@NonNull`, which is great !

But you could also use them here.
```suggestion
  @NonNull
  @SerializedName("lao")
  private final String laoId;

  @NonNull
  @SerializedName("election")
  private final String electionId;
```

File: fe2-android/app/src/main/java/com/github/dedis/popstellar/model/network/method/message/data/election/OpenElection.java [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+    return Action.OPEN.getAction();
+  }
+
+  public String getLaoId() {
+    return laoId;
+  }
+
+  public String getElectionId() {
+    return electionId;
+  }
```
Comment (GabrielFleischer) [inline fe2-android/app/src/main/java/com/github/dedis/popstellar/model/network/method/message/data/election/OpenElection.java:46]: And here

File: fe2-android/app/src/main/java/com/github/dedis/popstellar/ui/detail/LaoDetailViewModel.java [hunk 6]
```diff
@@ -259,6 +261,45 @@ protected void onCleared() {
...
+            .getMessageSender()
+            .publish(keyManager.getMainKeyPair(), channel, openElection)
+            .subscribe(
+                () -> {
+                  Log.d(TAG, "opened election successfully");
+                  //block action button on expandableListViewAdapter
+                  openElectionEvent();
+                },
+                error ->
+                    //modify the string after
```
Comment (GabrielFleischer) [inline fe2-android/app/src/main/java/com/github/dedis/popstellar/ui/detail/LaoDetailViewModel.java:297]: What does this comment mean ? I don't understand

File: fe2-android/app/src/main/java/com/github/dedis/popstellar/ui/detail/LaoDetailViewModel.java [hunk 7]
```diff
@@ -259,6 +261,45 @@ protected void onCleared() {
...
+    Lao lao = getCurrentLaoValue();
+    if (lao == null) {
+      Log.d(TAG, LAO_FAILURE_MESSAGE);
+      return;
+    }
+
+    Channel channel = e.getChannel();
+    String laoId = lao.getId();
+
+    //The time will have to be modified on the backend
```
Comment (GabrielFleischer) [inline fe2-android/app/src/main/java/com/github/dedis/popstellar/ui/detail/LaoDetailViewModel.java:281]: I don't think you formatted your code with the plugin.

There should be a space after the `//`.

File: fe2-android/app/src/main/java/com/github/dedis/popstellar/ui/detail/LaoDetailViewModel.java [hunk 8]
```diff
@@ -677,6 +718,8 @@ public void removeWitness(String witness) {
     return mOpenLaoDetailEvent;
   }
 
+  public LiveData<SingleEvent<Boolean>> getOpenElectionEvent(){return mOpenElectionEvent;}
+
```
Comment (GabrielFleischer) [inline fe2-android/app/src/main/java/com/github/dedis/popstellar/ui/detail/LaoDetailViewModel.java:722]: This also has a bad formatting.

File: fe2-android/app/src/main/java/com/github/dedis/popstellar/ui/detail/event/consensus/ElectionStartFragment.java [hunk 9]
```diff
@@ -180,9 +182,10 @@ private void setupButtonListeners(
       LaoDetailViewModel mLaoDetailViewModel,
       String electionId) {
     electionStart.setOnClickListener(
-        clicked ->
+        clicked -> {
             mLaoDetailViewModel.sendConsensusElect(
-                Instant.now().getEpochSecond(), electionId, "election", "state", "started"));
+                Instant.now().getEpochSecond(), electionId, "election", "state", "started");
+        });
```
Comment (GabrielFleischer) [inline fe2-android/app/src/main/java/com/github/dedis/popstellar/ui/detail/event/consensus/ElectionStartFragment.java:188]: Why is that modified ?

File: fe2-android/app/src/main/java/com/github/dedis/popstellar/ui/detail/event/consensus/ElectionStartFragment.java [hunk 10]
```diff
@@ -54,6 +54,8 @@
   private Button electionStart;
   private TextView electionStatus;
 
+  private Button electionStartNow;
+
```
Comment (GabrielFleischer) [inline fe2-android/app/src/main/java/com/github/dedis/popstellar/ui/detail/event/consensus/ElectionStartFragment.java:58]: Is this really used ?

File: fe2-android/app/src/main/java/com/github/dedis/popstellar/utility/handler/data/ElectionHandler.java [hunk 11]
```diff
@@ -92,6 +96,33 @@ public static void handleElectionResult(HandlerContext context, ElectionResult e
...
+
+    //If created --> open it
+    if (election.getState() == CREATED) {
+      election.setEventState(OPENED);
+    }
+
+    //Sets the start time to now
+    election.setStart(Instant.now().getEpochSecond());
+    Log.d(TAG, "election opened " + election.getStartTimestamp());
+    //election.setEventState(OPENED);
```
Comment (GabrielFleischer) [inline fe2-android/app/src/main/java/com/github/dedis/popstellar/utility/handler/data/ElectionHandler.java:121]: It is a bad practice to keep commented code in the main codebase.

File: fe2-android/app/src/test/java/com/github/dedis/popstellar/model/network/method/message/data/election/OpenElectionTest.java [hunk 12]
```diff
@@ -0,0 +1,60 @@
...
+
+    String random = "random";
+    assertNotEquals(openElection, new OpenElection(random, electionId, openedAt));
+    assertNotEquals(openElection, new OpenElection(laoId, random, openedAt));
+    assertNotEquals(openElection, new OpenElection(laoId, electionId, 0));
+
+    assertEquals(false, openElection.equals(null));
+
+    assertEquals(openElection, openElection);
+
```
Comment (GabrielFleischer) [inline fe2-android/app/src/test/java/com/github/dedis/popstellar/model/network/method/message/data/election/OpenElectionTest.java:48]: This empty line is not needed.

File: fe2-android/app/src/test/java/com/github/dedis/popstellar/utility/handler/ElectionHandlerTest.java [hunk 13]
```diff
@@ -60,6 +62,9 @@
 
   private static final Gson GSON = JsonModule.provideGson(DataRegistryModule.provideDataRegistry());
 
+  private static final long openedAt = 1633099883;
+
+
```
Comment (GabrielFleischer) [inline fe2-android/app/src/test/java/com/github/dedis/popstellar/utility/handler/ElectionHandlerTest.java:67]: There is too much space here.

File: protocol/examples/messageData/election_open.json [hunk 14]
```diff
@@ -4,4 +4,4 @@
     "lao": "fzJSZjKf-2cbXH7kds9H8NORuuFIRLkevJlN7qQemjo=",
     "election": "zG1olgFZwA0m3mLyUqeOqrG0MbjtfqShkyZ6hlyx1tg=",
     "opened_at": 1633099883
-}
\ No newline at end of file
+}
```
Comment (GabrielFleischer) [inline protocol/examples/messageData/election_open.json]: That is strange.

**Review by GabrielFleischer — CHANGES_REQUESTED** [review]
This is great ! 👍

There are some small things I do not completely agree with.

Please make sure you are using the `google-java-format` as it greatly help (does everything for you) when it comes to formatting.

Also make sure to fix the code smells raised by sonar cloud.

File: fe2-android/app/src/main/java/com/github/dedis/popstellar/model/network/method/message/data/election/OpenElection.java (hunk 5 above)
Comment (GabrielFleischer) [reply fe2-android/app/src/main/java/com/github/dedis/popstellar/model/network/method/message/data/election/OpenElection.java:45]: You can annotate `getLaoId()` and `getElectionId()` if you really want to use the NonNull annotation to the fullest.
But this is nitpicking.

File: fe2-android/app/src/main/java/com/github/dedis/popstellar/ui/detail/LaoDetailViewModel.java [hunk 15]
```diff
@@ -936,6 +980,13 @@ public void openSocialMedia() {
     mOpenSocialMediaEvent.setValue(new SingleEvent<>(true));
   }
 
+  /**
+   * Propagates the open election event
+   */
```
Comment (GabrielFleischer) [inline fe2-android/app/src/main/java/com/github/dedis/popstellar/ui/detail/LaoDetailViewModel.java:985]: The formatting does not follow the project-defined style.

**Review by GabrielFleischer — CHANGES_REQUESTED** [review]
I might have been to fast, but there are still some comment that were not addressed.

File: fe2-android/app/src/main/java/com/github/dedis/popstellar/utility/handler/data/ElectionHandler.java [hunk 16]
```diff
@@ -92,6 +94,32 @@ public static void handleElectionResult(HandlerContext context, ElectionResult e
...
+  public static void handleOpenElection(HandlerContext context, OpenElection openElection)
+      throws DataHandlingException {
+    LAORepository laoRepository = context.getLaoRepository();
+    Channel channel = context.getChannel();
+
+    Log.d(TAG, "handleOpenElection: channel " + channel);
+    Lao lao = laoRepository.getLaoByChannel(channel);
+    Election election = laoRepository.getElectionByChannel(channel);
+
+    if (election.getState() != CREATED) {
```
Comment (jbsv) [inline fe2-android/app/src/main/java/com/github/dedis/popstellar/utility/handler/data/ElectionHandler.java]: I think this should be set in line 61 during the setup, instead of: 
election.setEventState(OPENED);
or where is the state set to CREATED ?

**Review by jbsv — CHANGES_REQUESTED** [review]

File: fe2-android/app/src/main/java/com/github/dedis/popstellar/utility/handler/data/ElectionHandler.java (hunk 16 above)
Comment (ahominal) [reply fe2-android/app/src/main/java/com/github/dedis/popstellar/utility/handler/data/ElectionHandler.java]: When the election Setup message is received, the election is set to CREATED. When it receives the OpenElection, it sets to OPENED. It made more sense to me.

**Review by GabrielFleischer — APPROVED** [review]
LGTM

File: fe2-android/app/src/main/java/com/github/dedis/popstellar/utility/handler/data/ElectionHandler.java (hunk 16 above)
Comment (jbsv) [reply fe2-android/app/src/main/java/com/github/dedis/popstellar/utility/handler/data/ElectionHandler.java]: That's also what I understood from your code, but then why is it set to OPENED on line 61 in handleElectionSetup() ? It looks like a bug to me. And by the way, the naming consistency of the functions would be better respected with handleElectionOpen().

**Comment by jbsv** [comment]
2 codes smells and a bug (I think) are left.

**Review by jbsv — CHANGES_REQUESTED** [review]

File: fe2-android/app/src/test/java/com/github/dedis/popstellar/utility/handler/ElectionHandlerTest.java [hunk 17]
```diff
@@ -199,4 +202,25 @@ public void testHandleElectionEnd() throws DataHandlingException {
...
+    MessageGeneral message = new MessageGeneral(SENDER_KEY, openElection, GSON);
+
+    for (EventState state : EventState.values()) {
+      election.setEventState(state);
+
+      if (state == EventState.CREATED) {
+
+        //If the previous state was CREATED, it should change it to OPENED.
+        messageHandler.handleMessage(laoRepository, messageSender,
+            election.getChannel(), message);
```
Comment (GabrielFleischer) [inline fe2-android/app/src/test/java/com/github/dedis/popstellar/utility/handler/ElectionHandlerTest.java:218]: Don't you want to send the message for each state and not only for created state ?

File: fe2-android/app/src/test/java/com/github/dedis/popstellar/utility/handler/ElectionHandlerTest.java [hunk 18]
```diff
@@ -199,4 +202,25 @@ public void testHandleElectionEnd() throws DataHandlingException {
...
+        assertEquals(EventState.OPENED, election.getState());
+        //Test for current TimeStamp
+        assertEquals(Instant.now().getEpochSecond(), election.getStartTimestamp());
```
Comment (GabrielFleischer) [inline fe2-android/app/src/test/java/com/github/dedis/popstellar/utility/handler/ElectionHandlerTest.java:221]: I am not sure, I understand. I think this should be equal to `openedAt`.

**Review by GabrielFleischer — CHANGES_REQUESTED** [review]
I found some oddities in the test while trying to understand it.

I think there is an issue with the start time of an opened election, but I am not certain of that.

File: fe2-android/app/src/test/java/com/github/dedis/popstellar/utility/handler/ElectionHandlerTest.java (hunk 17 above)
Comment (ahominal) [reply fe2-android/app/src/test/java/com/github/dedis/popstellar/utility/handler/ElectionHandlerTest.java:218]: It should transition from CREATED --> OPENED only when the current state is CREATED, this way I need to check for other states as well

File: fe2-android/app/src/test/java/com/github/dedis/popstellar/utility/handler/ElectionHandlerTest.java (hunk 18 above)
Comment (ahominal) [reply fe2-android/app/src/test/java/com/github/dedis/popstellar/utility/handler/ElectionHandlerTest.java:221]: When it opens, it should it modify the start time of the election (the current time). This is is the way it's implemented on the backend, so I made the same for java.

File: fe2-android/app/src/main/java/com/github/dedis/popstellar/utility/handler/data/ElectionHandler.java (hunk 16 above)
Comment (ahominal) [reply fe2-android/app/src/main/java/com/github/dedis/popstellar/utility/handler/data/ElectionHandler.java]: It should be CREATED on the line 61, at least it is on local. Strange it didn't push. Thanks for clarifying.

**Review by jbsv — DISMISSED** [review]
Sorry, github was acting against me and I didn't see the changes in the setup handler

File: fe2-android/app/src/test/java/com/github/dedis/popstellar/utility/handler/ElectionHandlerTest.java (hunk 17 above)
Comment (GabrielFleischer) [reply fe2-android/app/src/test/java/com/github/dedis/popstellar/utility/handler/ElectionHandlerTest.java:218]: I think you should move the line `messageHandler.handleMessage(laoRepository, messageSender, election.getChannel(), message);` should be moved outside the loop.

So the test makes sure that the election isn't opened when the state is not create.

File: fe2-android/app/src/test/java/com/github/dedis/popstellar/utility/handler/ElectionHandlerTest.java (hunk 18 above)
Comment (GabrielFleischer) [reply fe2-android/app/src/test/java/com/github/dedis/popstellar/utility/handler/ElectionHandlerTest.java:221]: This is an odd way to do it, and it might be a security issue.

Also, why does the open election data has an `opened_at` field then ?

I was not able to find the documentation on that specific message, but could you double-check with your team that it is the way it should be done ?

**Review by GabrielFleischer — APPROVED** [review]
LGTM
